framework/src/userobject/UserObjectInterface.C

```diff
@@ -7,34 +7,86 @@
+                 ? _uoi_moose_object.parameters().get<THREAD_ID>("_tid")
+                 : 0)
 {
 }
 
 const UserObject &
-UserObjectInterface::getUserObjectBase(const std::string & name) const
+UserObjectInterface::getUserObjectBase(const std::string & param_name) const
 {
-  return _uoi_feproblem.getUserObjectBase(_uoi_params.get<UserObjectName>(name));
+  const auto & params = _uoi_moose_object.parameters();
+
+  if (!params.isParamValid(param_name))
+    _uoi_moose_object.mooseError("Failed to get a parameter with the name \"",
+                                 param_name,
+                                 "\".",
+                                 "\n\nKnown parameters:\n",
+                                 _uoi_moose_object.parameters());
+
+  // Other interfaces will use this interface (PostprocessorInterface, VectorPostprocessorInterface)
+  // to grab UOs with a specialized name, so we need to check them all
+  std::string object_name;
+  if (params.have_parameter<UserObjectName>(param_name))
+    object_name = params.get<UserObjectName>(param_name);
+  else if (params.have_parameter<PostprocessorName>(param_name))
+    object_name = params.get<PostprocessorName>(param_name);
+  else if (params.have_parameter<VectorPostprocessorName>(param_name))
+    object_name = params.get<VectorPostprocessorName>(param_name);
+  else if (params.have_parameter<std::string>(param_name))
+    object_name = params.get<std::string>(param_name);
+  else
+    _uoi_moose_object.paramError(param_name,
+                                 "Parameter of type \"",
+                                 params.type(param_name),
+                                 "\" is not an expected type for getting a UserObject");
+
+  if (!_uoi_feproblem.hasUserObject(object_name))
+    _uoi_moose_object.paramError(
+        param_name, "The requested object with the name \"", object_name, "\" was not found.");
+
+  return getUserObjectBaseByName(object_name);
 }
 
 const UserObject &
-UserObjectInterface::getUserObjectBaseByName(const std::string & name) const
+UserObjectInterface::getUserObjectBaseByName(const UserObjectName & object_name) const
 {
-  return _uoi_feproblem.getUserObjectBase(name);
+  if (!_uoi_feproblem.hasUserObject(object_name))
+    _uoi_moose_object.mooseError(
+        "The requested object with the name \"", object_name, "\" was not found.");
+
+  const auto & uo_base_tid0 = _uoi_feproblem.getUserObjectBase(object_name, /* tid = */ 0);
+  const THREAD_ID tid = needThreadedCopy(uo_base_tid0) ? _uoi_tid : 0;
+  return _uoi_feproblem.getUserObjectBase(object_name, tid);
 }
 
 bool
 UserObjectInterface::needThreadedCopy(const UserObject & uo) const
 {
-  return (dynamic_cast<const DiscreteElementUserObject *>(&uo) != NULL) ||
-         (dynamic_cast<const ThreadedGeneralUserObject *>(&uo) != NULL);
+  return dynamic_cast<const DiscreteElementUserObject *>(&uo) ||
+         dynamic_cast<const ThreadedGeneralUserObject *>(&uo);
```

Comment:
I know the implementation of this function isn't being changed here - but it seems to me like this function should return true if `uo` has threaded copies - and there are several user object types for which this is the case that aren't checked here - i.e. ElementUserObject subclasses, etc.  Does this need to be fixed?

Reply:
I actually thought about putting something like
```
virtual threaded() const { return false; }
```
in UserObject, and overriding it where needed and just using that instead. Thoughts?

Reply:
Yeah - that would be better than the casting nonsense.